tests/integration/src/xvm.rs

```diff
@@ -0,0 +1,334 @@
+
+/* Call EVM Payable:
+
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+
+use ink::env::{DefaultEnvironment, Environment};
+use ink::prelude::vec::Vec;
+
+#[ink::contract(env = CustomEnvironment)]
+mod call_xvm_payable {
+    use super::*;
+
+    #[ink(storage)]
+    pub struct CallXvmPayable {}
+
+    impl CallXvmPayable {
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            Self {}
+        }
+
+        #[ink(message, payable, selector = 42)]
+        pub fn call_xvm_payable(
+            &self,
+            target: Vec<u8>,
+            input: Vec<u8>,
+        ) -> CallResult {
+            let value = Self::env().transferred_value();
+            // Calling EVM
+            Self::env().extension().call(0x0F, target, input, value)
+        }
+    }
+}
+
+pub type CallResult = u32;
+
+#[ink::chain_extension]
+pub trait XvmCall {
+    type ErrorCode = u32;
+
+    #[ink(extension = 0x00010001, handle_status = false)]
+    fn call(vm_id: u8, target: Vec<u8>, input: Vec<u8>, value: u128) -> CallResult;
+}
+
+pub enum CustomEnvironment {}
+impl Environment for CustomEnvironment {
+    const MAX_EVENT_TOPICS: usize = <DefaultEnvironment as Environment>::MAX_EVENT_TOPICS;
+
+    type AccountId = <DefaultEnvironment as Environment>::AccountId;
+    type Balance = <DefaultEnvironment as Environment>::Balance;
+    type Hash = <DefaultEnvironment as Environment>::Hash;
+    type BlockNumber = <DefaultEnvironment as Environment>::BlockNumber;
+    type Timestamp = <DefaultEnvironment as Environment>::Timestamp;
+
+    type ChainExtension = XvmCall;
+}
+
+ */
+
+#[test]
+fn evm_payable_call_via_xvm_works() {
+    new_test_ext().execute_with(|| {
+        let evm_payable_addr = deploy_evm_contract(EVM_PAYABLE);
+
+        let value = UNIT;
+        assert_ok!(Xvm::call(
+            Context {
+                source_vm_id: VmId::Wasm,
+                weight_limit: Weight::from_parts(1_000_000_000, 1024 * 1024),
+            },
+            VmId::Evm,
+            ALICE,
+            evm_payable_addr.as_ref().to_vec(),
+            // Calling `deposit`
+            hex::decode("d0e30db0").expect("invalid selector hex"),
+            value,
+        ));
+        assert_eq!(
+            Balances::free_balance(account_id_from(evm_payable_addr)),
+            value
+        );
+
+        assert_ok!(Xvm::call(
+            Context {
+                source_vm_id: VmId::Wasm,
+                weight_limit: Weight::from_parts(10_000_000_000, 1024 * 1024),
+            },
+            VmId::Evm,
+            ALICE,
+            evm_payable_addr.as_ref().to_vec(),
+            // `Calling withdraw`
+            hex::decode("3ccfd60b").expect("invalid selector hex"),
+            0,
+        ));
+        assert_eq!(
+            Balances::free_balance(account_id_from(evm_payable_addr)),
+            ExistentialDeposit::get(),
+        );
+    });
+}
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The test function `evm_payable_call_via_xvm_works` tests the cross-VM call from the Wasm VM to the EVM VM. The test deploys an EVM contract, performs a payable call to the contract's `deposit` function, and verifies the balance change. No issues found.



<!-- This is an auto-generated comment by OSS CodeRabbit -->